Support GitHub Enterprise

jankeromnes opened the pull request:
## Description
<!-- Describe your changes in detail -->

- [x] Fix raw file URL for GitHub Enterprise to allow fetching `.gitpod.yml` files successfully
- [x] Fix the `Grant Access` button/flow for private repositories on GitHub Enterprise
- [x] Allow creating Projects and running Prebuilds for GitHub Enterprise repositories
- [x] Drive-by: Trim `clientId` and `clientSecret` values when creating a new Integration
- [x] Drive-by: In dev-staging, don't truncate the callbackURL of new Integrations

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes https://github.com/gitpod-io/gitpod/issues/8501 https://github.com/gitpod-io/gitpod/issues/8429

## How to test
<!-- Provide steps to test this PR -->

1. Have a GitHub Enterprise server account (for example, on https://ghe.gitpod-self-hosted.com )
2. Have a public repository there (e.g. https://ghe.gitpod-self-hosted.com/roboquat/public-repo) and a private repository (e.g. https://ghe.gitpod-self-hosted.com/roboquat/top-secret)
3. Log in to https://jx-fix-ghe-yml.staging.gitpod-dev.com/integrations (e.g. using your github.com account)
4. Follow https://www.gitpod.io/docs/github-integration#github-enterprise-server to create a new Integration for your GitHub Enterprise server, and activate it
5. Open your public repository from 2. in https://jx-fix-ghe-yml.staging.gitpod-dev.com/open -- it should work
6. Open your private repository from 2. -- it should ask you to `Grant Access`, then also work
7. Add your two repositories as Projects in https://jx-fix-ghe-yml.staging.gitpod-dev.com/new
8. Once the Projects are created, a first Prebuild should run and complete successfully
9. Push a new commit to one of your repositories -- the new commit should be detected by Gitpod, and trigger a new Prebuild

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
Support GitHub Enterprise
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

jankeromnes commented:
Taking this back to draft: Should work fine as is, but needs a GitHub Enterprise integration with the preview environment to test it, and once I have that, I might as well add in a few more fixes.

andrew-farries commented on components/server/ee/src/prebuilds/github-service.ts:
```diff
@@ -0,0 +1,52 @@
+    @inject(GitHubGraphQlEndpoint) protected readonly githubQueryApi: GitHubGraphQlEndpoint;
+    @inject(GitHubRestApi) protected readonly githubApi: GitHubRestApi;
+    @inject(Config) protected readonly config: Config;
+    @inject(AuthProviderParams) protected authProviderConfig: AuthProviderParams;
+	@inject(TokenService) protected tokenService: TokenService;
+	@inject(GithubContextParser) protected githubContextParser: GithubContextParser;
```
Looks like some wonky formatting here.

jankeromnes replied:
Who put _tabs_ into my soup? 😳 Fixed ✅

andrew-farries commented on components/server/src/bitbucket-server/bitbucket-server-repository-provider.ts:
```diff
@@ -102,6 +102,11 @@ export class BitbucketServerRepositoryProvider implements RepositoryProvider {
         return [];
     }
 
+    async getProviderRepos(user: User): Promise<ProviderRepository[]> {
+        // FIXME(janx): Not implemented yet
+        return [];
+    }
+
```
Assume this FIXME is not intended to be addressed in this PR?

jankeromnes replied:
You're right. 👍

As discussed earlier, I'd love to re-unify these "repository-handling" classes soon:
- [RepositoryProvider](https://github.com/gitpod-io/gitpod/blob/main/components/server/src/repohost/repository-provider.ts)
- [RepositoryService](https://github.com/gitpod-io/gitpod/blob/main/components/server/src/repohost/repo-service.ts)
- {GitHub,GitLab,Bitbucket}AppSupport (single method, no parent class, see e.g. [GitLabAppSupport](https://github.com/gitpod-io/gitpod/blob/main/components/server/ee/src/gitlab/gitlab-app-support.ts))

An in fact, this method `getProviderRepos` is actually already implemented in [BitbucketAppSupport](https://github.com/gitpod-io/gitpod/blob/main/components/server/ee/src/bitbucket/bitbucket-app-support.ts#L18-L78).

But I'd prefer leaving this refactoring for a follow-up Pull Request, so adding a FIXME for now.

jankeromnes replied:
Ah, given that the EE vs not-EE split might still be relevant (due to licensing concerns), maybe it does make sense to keep two seperate classes (for example, unify the code into `RepositoryProvider` in not-EE and `RepositoryService` in EE -- in that case I think we'd need to move this `getProviderRepos` method into `RepositoryService`, which sort of makes sense because "get webhook-compatible repos" and "is repo webhook-compatible" and "install webhook on repo" are all related)

jankeromnes replied:
Done.

andrew-farries commented on components/server/src/bitbucket/bitbucket-repository-provider.ts:
```diff
@@ -104,6 +104,11 @@ export class BitbucketRepositoryProvider implements RepositoryProvider {
         return [];
     }
 
+    async getProviderRepos(user: User): Promise<ProviderRepository[]> {
+        // FIXME(janx): Not implemented yet
+        return [];
+    }
+
```
Same question about the FIXME.

andrew-farries commented on components/server/src/gitlab/gitlab-repository-provider.ts:
```diff
@@ -100,6 +100,11 @@ export class GitlabRepositoryProvider implements RepositoryProvider {
         return [];
     }
 
+    async getProviderRepos(user: User): Promise<ProviderRepository[]> {
+        // FIXME(janx): Not implemented yet
+        return [];
+    }
+
```
Same question about the FIXME

andrew-farries reviewed (commented):
We went through some of this on the call earlier, but could the 'how to test' section be filled in so that I can test an integration with GHE using our installation at https://ghe.gitpod-self-hosted.com/ ?

andrew-farries commented:
Thanks for adding the testing steps. 

When I push a new commit to https://ghe.gitpod-self-hosted.com/roboquat/public-repo, I see the commit in the `Branches` tab but I don't get a corresponding prebuild on the `Prebuilds` tab:

https://user-images.githubusercontent.com/8225907/157419757-810d4e12-ffbf-460d-b38c-dc646cc1e7b6.mov

jankeromnes commented:
Awesome, many thanks for testing @andrew-farries! 🙏

Indeed, the triggering of new Prebuilds on pushes is currently broken, as I'm still adjusting the webhook signature verification. I'm currently iterating on it like so:
- pushing a new commit to the test repo
- observing the server logs for any errors
- fixing errors, then waiting for a new deployment of this PR

Happy to have a chat about where things are. 🙂 I'm hopeful that it will be working and ready for review quite soon 🤞 (at which point we'll be able to take the PR out of Draft).

jankeromnes commented:
Thanks @andrew-farries for helping discuss, review, and test this. 💯

It now works as expected, and should be about ready for review. 🚀

jankeromnes commented:
Holding because of the minor TODOs above:

/hold

jankeromnes reviewed (commented):
(Forgot to actually submit the minor TODO comments)

geropl commented on components/server/ee/src/prebuilds/github-enterprise-app.ts:
```diff
@@ -0,0 +1,195 @@
+     *
+     * First of all it tries to find the project for the given `cloneURL`, then it tries to
+     * find the installer, which is also supposed to be a team member. As a fallback, it
+     * looks for a team member which also has a bitbucket.org connection.
```
> bitbucket.org

Copy'n'paste

geropl commented:
@jankeromnes Thank you for tackling this, and providing a working solution this fast! :100: :pray: 

We discussed this earlier, just want to point it out again:
 - This PR breaks with the existing pattern we use for integrations `type == 'GitHub'`, where we install GitHub Apps on the org/repo level. Instead, it uses the OAuth-app (user-impersonation) to install the webhook (see ["Differences between GitHub Apps and OAuth Apps"](https://docs.github.com/en/developers/apps/getting-started-with-apps/differences-between-github-apps-and-oauth-apps)).
 - I'm fine with deviating for GitHub Enterprise, especially if it's temporary for the sake of progress. But then someone must own the responsibility to make this coherent again a) technically (by implementing a any required changes), and b) product-wise (by re-discovering the differences between both approaches). (/cc @jldec )
 - I want to
   a) ensure we don't take up additional long-term technical debt by introducing new integration modes
   b) make sure someone (not me :upside_down_face: ) follows up on this, because I see no value being added to our goals in this move.

If I'm missing something here - be it technical, or other - please let me know! Also, really don't want to hinder the good progress here, just want to make sure by point is clear. :slightly_smiling_face:

geropl commented:
On the technical side: How do we ensure we have the right scopes for:
 - registering webhooks (`write:repo_hook` or `write:org_hook` scope, [ref](https://docs.github.com/en/developers/apps/getting-started-with-apps/differences-between-github-apps-and-oauth-apps#requesting-permission-levels-for-resources))
 - updating Commit statuses (`repo:status` scope, [ref](https://docs.github.com/en/developers/apps/getting-started-with-apps/differences-between-github-apps-and-oauth-apps#requesting-permission-levels-for-resources))

jankeromnes commented:
Thanks @geropl for the quick review! 🙏

>  - This PR breaks with the existing pattern we use for integrations `type == 'GitHub'`, where we install GitHub Apps on the org/repo level. Instead, it uses the OAuth-app (user-impersonation) to install the webhook (see ["Differences between GitHub Apps and OAuth Apps"](https://docs.github.com/en/developers/apps/getting-started-with-apps/differences-between-github-apps-and-oauth-apps)).

I don't think it "breaks the pattern", because the GitHub App is [a singleton](https://github.com/gitpod-io/gitpod/blob/54a765d321236ef42dceeb154cf477b17a8aaf32/components/server/src/config.ts#L72-L80) in Gitpod (i.e. you can only have one GitHub App per Gitpod installation), while we already allow an arbitrary number of Integrations (of type `GitHub` or `GitLab`) which are already OAuth apps. As we've seen, there are already 13 GitHub Enterprise integrations on gitpod.io. This Pull Request makes them actually work.

Also, webhooks and the GitHub App are not mutually exclusive. As a next step to this PR, I'd love to explore what it could look like if you have a single Gitpod Self-Hosted installation, and connect its GitHub App singleton to a single GitHub Enterprise server (as opposed to connecting it to github.com as we do on SaaS). Ironing out any bugs in that process could create a nice alternative integration point specifically for Self-Hosted / GHE.

> - I want to
>   a) ensure we don't take up additional long-term technical debt by introducing new integration modes

- We already allow arbitrary OAuth App integrations for GitLab and GitHub Enterprise in Gitpod (that's how the custom Integrations in `/integrations` work)
- This Pull Request just makes the existing GitHub Enterprise integrations work properly, by using the same mechanism as GitLab and Bitbucket (webhooks)
- In the future, I would love to improve these integrations (GitLab, Bitbucket and GitHub Enterprise) to make them post visible status updates on repositories, commits and Pull Requests, to make GitLab Bitbucket and GitHub Enterprise reach parity with our current `github.com` support

>   b) make sure someone (not me 🙃 ) follows up on this, because I see no value being added to our goals in this move.

As mentioned above, I'm happy to follow up, e.g. by testing a Gitpod Self-Hosted using a GitHub App connected to a GitHub Enterprise server. As mentioned above, I don't think webhooks and the GitHub App are mutually exclusive.

jankeromnes commented:
> On the technical side: How do we ensure we have the right scopes for:
>  - registering webhooks (`write:repo_hook` or `write:org_hook` scope, [ref](https://docs.github.com/en/developers/apps/getting-started-with-apps/differences-between-github-apps-and-oauth-apps#requesting-permission-levels-for-resources))
>  - updating Commit statuses (`repo:status` scope, [ref](https://docs.github.com/en/developers/apps/getting-started-with-apps/differences-between-github-apps-and-oauth-apps#requesting-permission-levels-for-resources))

I believe these scopes are all included in the `repo` (or `public_repo`) scopes we already request (see mentions of "webhooks" and "commit status" in [available scopes](https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps#available-scopes))

jankeromnes commented:
For some reason Werft decided to hide the preview URL of this Pull Request, but the preview itself is still live FYI:

https://jx-fix-ghe-yml.staging.gitpod-dev.com/

AlexTugarev commented on components/server/ee/src/prebuilds/github-service.ts:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    async installAutomatedPrebuilds(user: User, cloneUrl: string): Promise<void> {
+        const { owner, repoName: repo } = await this.githubContextParser.parseURL(user, cloneUrl);
+        const webhooks = (await this.githubApi.run(user, gh => gh.repos.listWebhooks({ owner, repo }))).data;
```
as long as they are bound in the container module which are loaded per host (see HostContextProvider) this should be fine, at least that exactly the purpose of host contexts.

jankeromnes replied:
Awesome, many thanks for this explanation @AlexTugarev. 🙏

AlexTugarev commented on components/server/ee/src/prebuilds/github-service.ts:
```diff
@@ -0,0 +1,60 @@
+        if (host === 'github.com' || host !== this.authProviderConfig.host) {
+            return false;
+        }
+        return true;
```
that's not enough. see `GitLabService.canInstallAutomatedPrebuilds`, we need to check if `user` has admin/maintainer permissions for this repo.

jankeromnes replied:
Added a GraphQL query to check that `viewerPermission === 'ADMIN'` (`'MAINTAIN'` is not sufficient -- [source](https://docs.github.com/en/organizations/managing-access-to-your-organizations-repositories/repository-roles-for-an-organization#permissions-for-each-role))

AlexTugarev commented on components/server/ee/src/prebuilds/github-enterprise-app.ts:
```diff
@@ -0,0 +1,197 @@
+            // Verify the webhook signature
+            const signature = req.header('X-Hub-Signature-256');
+            const body = (req as any).rawBody;
+            const tokenEntries = (await this.userDB.findTokensForIdentity(gitpodIdentity)).filter(tokenEntry => {
```
how many `tokenEntry`s are expected to be there? it would be great to pick the current one if there are multiple.

jankeromnes replied:
As many `tokenEntry`s as repositories you've installed prebuilds on.

I didn't filter on the `cloneUrl` scope we also add to tokens, because we don't update the `cloneUrl`s in tokens when repositories get renamed.

AlexTugarev replied:
oh! that sounds really expensive. if we get many of them, such operations will drive the event loop lag even higher.

jankeromnes commented on components/dashboard/src/projects/NewProject.tsx:
```diff
@@ -525,17 +525,14 @@ function GitProviders(props: {
                     Select a Git provider first and continue with your repositories.
                 </div>
                 <div className="mt-6 flex flex-col space-y-3 items-center pb-8">
-                    {filteredProviders().map(ap => {
```
Good catch!

I've removed this filter from the UI, because I consider that:
- Your unsupported provider is visible in the UI, but it might potentially fail somewhere down the line because we haven't fully tested it

is a better UX than:
- Your unsupported provider is not visible in the UI, because we haven't tested it yet

In other words, maybe it's better to try and give users more agency here. If things work out of the box, great! If they break in some specific way, we can file an issue and try to solve it quickly. (But if we completely hide these, we're sort of ensuring that they'll never work.)

AlexTugarev replied:
> is a better UX than:

really? 

how is a (untested ===) broken state better than hiding untested paths? 

> But if we completely hide these, we're sort of ensuring that they'll never work.

We got a lot of requests and scheduled several tasks to extend the support in this area in the past year, and this PR is an example for it. So this isn't true.

jankeromnes replied:
I'm still biased towards leaving all doors open and all panels unlocked, so that savvy users can help us test and find their way around / fix their own problems, as opposed to shutting every door and having just one supported path.

However, I also understand your point of view, so I'll re-instate the UI filter I removed (and adjust it to show GitHub Enterprise as well).

jankeromnes commented on components/server/src/server.ts:
```diff
@@ -117,9 +117,9 @@ export class Server<C extends GitpodClient, S extends GitpodServer> {
-        app.use(bodyParser.json())
-        app.use(bodyParser.urlencoded({ extended: true }))
+        // Read bodies as JSON (but keep the raw body just in case)
+        app.use(bodyParser.json({ verify: (req, res, buffer) => { (req as any).rawBody = buffer; }}));
```
You're right, with this change we'll carry the raw body of JSON requests we receive. I'm assuming that, apart from some webhooks, we don't receive too many / very large JSON payloads.

AlexTugarev commented on components/server/ee/src/prebuilds/github-service.ts:
```diff
@@ -0,0 +1,60 @@
+                await this.githubApi.run(user, gh => gh.repos.deleteWebhook({ owner, repo, hook_id: webhook.id }));
+            }
+        }
+        // TODO(janx): Also delete old tokens with scopes `GitHubService.PREBUILD_TOKEN_SCOPE` and `cloneUrl`?
```
please, yes! unless it's already done in `createGitpodToken` op.

AlexTugarev replied:
isn't reusing a token another option here?
just in case, I'm thinking of the `db-sync` dilemma.

jankeromnes replied:
> unless it's already done in `createGitpodToken` op.

Ah, good thought! How nice that this is actually the case: 😄

https://github.com/gitpod-io/gitpod/blob/3bd683e6eaf391deeb3b3b7803d7fe4d4e513d9d/components/server/src/user/token-service.ts#L64-L67

> isn't reusing a token another option here?
> just in case, I'm thinking of the `db-sync` dilemma.

Could you please explain the `db-sync` dilemma regarding whether to reuse a token vs creating a new one? 🤔

FYI, I followed the same pattern as GitLab and Bitbucket webhooks, i.e. generate a new token for every repo (with the scopes `"prebuilds"` and `cloneUrl`).

AlexTugarev replied:
> Could you please explain the db-sync dilemma regarding whether to reuse a token vs creating a new one? 🤔

If you override tokens in the DB for a particular repo, they won't be available in the other cluster until the following db-sync cycle. But that might be a general problem with self-managed webhooks.

jankeromnes commented on components/server/src/github/api.ts:
```diff
@@ -174,7 +174,7 @@ export class GitHubRestApi {
             const response = (await operation(userApi));
             const statusCode = response.status;
-            if (statusCode !== 200) {
+            if (!(statusCode >= 200 && statusCode < 300)) {
```
There are actually a bunch of different HTTP success codes: [Wikipedia](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#2xx_success)

And indeed, when you create a webhook using the GitHub API, it responds with `201 Created`, which is obviously not an error. 😇

jankeromnes commented on components/server/src/github/github-repository-provider.ts:
```diff
@@ -129,6 +129,20 @@ export class GithubRepositoryProvider implements RepositoryProvider {
         return (result.data.viewer?.repositoriesContributedTo?.edges || []).map((edge: any) => edge.node.url)
     }
 
+    async getProviderRepos(user: User): Promise<ProviderRepository[]> {
```
Refactored `RepositoryProvider.getProviderRepos` to `RepositoryService.getRepositoriesForAutomatedPrebuilds`.

Resolving here, please feel free to re-open a conversation if the new name still isn't right. 😁

geropl commented:
@jankeromnes Thanks for the [explanation](https://github.com/gitpod-io/gitpod/pull/8574#issuecomment-1062938200).

> As we've seen, there are already 13 GitHub Enterprise integrations on gitpod.io. This Pull Request makes them actually work.

:bulb: It's news to me that we allow/already support these. I saw you slack post earlier this week, but thought the entries where due to bugs.

> I don't think it "breaks the pattern"

It's bit complicated because of the two use cases:
 - qualified, by admin, at config-time: connect 1 integration with a single, statically configured GitHub app
 - unqualified, by user, at runtime: connect n integrations; so far we only had GitLab/BitBucket, so used webhooks (and now for GHE as well)

My main concern was that I understood that we want to achieve 1.), but this PR uses the techniques we employ for 2.). This is what I meant with "breaking the pattern". Instead of focusing on the problem at hand (having a customer configure they own GitHub App for their own installation) we seem to be fixing the GHE integration in general. Which is also an interesting goal, but feels a bit out of scope here.

> Also, webhooks and the GitHub App are not mutually exclusive

Not at runtime, yes, but I'd really like us to see it this way, because it additional maintenance burden, and now we have to schedule the task of cleaning it up (e.g., deciding for one way or the other).

> As a next step to this PR, I'd love to explore what it could look like if you have a single Gitpod Self-Hosted installation, and connect its GitHub App singleton to a single GitHub Enterprise server (as opposed to connecting it to github.com as we do on SaaS)

Awesome, really looking forward to this. :muscle:

jankeromnes commented:
Many thanks @andrew-farries, @geropl and @AlexTugarev for your super helpful reviews! 🙏

I've addressed all nits, so this seems ready to be merged. 🚢

/unhold

andrew-farries commented on components/server/ee/src/prebuilds/github-enterprise-app.ts:
```diff
@@ -0,0 +1,197 @@
+                const sig = 'sha256=' + createHmac('sha256', user.id + '|' + tokenEntry.token.value)
+                    .update(body)
+                    .digest('hex');
+                return sig === signature;
```
This is probably pedantry at this stage, but comparing the HMAC signatures using a simple string comparison is considered insecure because of the potential for timing attacks. 

See [the github webhook docs](https://docs.github.com/en/developers/webhooks-and-events/webhooks/securing-your-webhooks). Seems the best way to do it is with `crypto.timingSafeEqual` as suggested [here](https://stackoverflow.com/questions/51486432/what-is-the-preferred-way-of-comparing-hmac-signatures-in-node).
